Remove static deps optimization

jeremiedimino opened the pull request:
This PR disables the optimisation to build dependencies known statically in parallel with the computation of the action and the full  set of dependencies, which simplified the code quite a bit.

I ran a benchmark and didn't observe a different, suggesting this optimisation is not worth it. The benchmark I ran is:

```
$ opam source patdiff.v0.14.0
$ cd patdiff.v0.14.0
$ opam monorepo lock
$ opam monorepo pull
$ time dune build --profile release @install
```

on a machine with 20 cores. The only difference is that the total number of jobs increases quicker with the main branch.

This patch currently breaks `dune external-lib-deps`.

rgrinberg commented:
> on a machine with 20 cores. The only difference is that the total number of jobs increases quicker with the main branch.

You might also want to benchmark a zero build. Though I think the result will be even more favorable for this PR. From my profiling of zero builds, quite a lot of time is spent on these set operations in `static_deps`.

> This patch currently breaks dune external-lib-deps.

Did we talk about formally breaking this for 3.0? It seems like a good time to remove this thing.

jeremiedimino commented:
> You might also want to benchmark a zero build. Though I think the result will be even more favorable for this PR. From my profiling of zero builds, quite a lot of time is spent on these set operations in static_deps.

Ah, interesting. I just tried, I'm seeing a lot of noise but there seems to be a small improvement on null builds with the same benchmark.

> Did we talk about formally removing this for 3.0? It seems like a good time to finally do it.

Yeah, that's a good point. At worse, we could recover it by other means if we don't want to delete it straight away.

jeremiedimino commented:
I tried a slight variation (last commit): I wait for the dependencies directly in `Build.exec`. The run time still seems the same, but we discover more of the dag upfront. i.e. the total number of jobs increases even faster than before.

However, it means that we can no longer memoise just the evaluation of the action, which is a problem for `package_deps`

jeremiedimino commented:
This PR is ready for review. (/cc @aalekseyev @snowleopard).

Now that we no longer need to support static analysis of rules, I was able to simply the `Build_request` module in `build_system.ml`. In particular, `execute_rules` now just calls `Action_builder.Make_exec().exec`, which does everything in a single pass:

- compute the result (waiting for files when necessary, such as for `Action_builder.contents`)
- compute the full set of action dependencies
- wait for all the dependencies to be built

A side effect of this is that we tend to scan more of the graph in case of errors. For instance, in the `formatting.t` test, without the `@fmt -> @lang2/partial/fmt` change the test reports diffs that it didn't reported before because the build was stopping earlier. This seems like a good change since reporting all errors is what dune tries to do by default.

The previous code used to have a memoised function that only evaluated a rule and reported the action and set of dependencies. This was used in `package_deps` and `evaluate_rule`. With the new code, we no longer have such a memoized function, so I adapted it as follow:

## package_deps

I changed `execute_rules` and `build_file` to return the set of dependencies computed. While doing so, I noticed that the current implementation of `package_deps` is broken in presence of dynamic actions (the ones using `dune-action-plugin`) as it doesn't scan dependencies discovered dynamically while the action is running. Given that, changing `execute_rule` to return the set of dependencies seems right.

`build_file` used to have a cutoff and adding a cutoff on the version that returns the set of dependencies doesn't seem right. So I split it in two:

- `build_file` which returns the set of dependencies and has no cutoff
- `build_file_unit` which calls `build_file`, ignores the result and has a cutoff

## evaluate_rules

This function is only used for the `dune rules` command. In this function:
- we need to execute some of the rules, for instance we need to call `ocamldep` to compute all the module dependencies
- for some of the rules we only need to compute the action
- for some rules,  we need both

With the old code, we could share computing the action between the code path executing the rule and the code path only computing the action. Keeping this sharing means keeping the staging for normal builds, so I didn't keep it. This means that when running `dune rules`, some actions will be computed twice, which seems acceptable to me as this command is only used for debugging.

snowleopard commented:
> `build_file` used to have a cutoff and adding a cutoff on the version that returns the set of dependencies doesn't seem right

@jeremiedimino I didn't understand the reasoning here. Why do you think adding a cutoff for this version isn't right?

snowleopard commented on src/dune_engine/action_builder.ml:
```diff
@@ -277,67 +277,6 @@ let progn ts =
-
-  let rec static_deps : type a. a t -> Static_deps.t =
-   fun t ->
-    let file_exists = Fdecl.get file_exists_fdecl in
```
Do we still use `file_exists` and the corresponding `Fdecl` and constructor anywhere? Perhaps, this can also be simplified away.

snowleopard replied:
Oh, we do use the constructor (missed on the first pass), but do we still need the `Fdecl`? Perhaps, we can now pass it directly.

jeremiedimino replied:
Good point. I added it to the functor.

snowleopard replied:
:+1:

snowleopard commented:
> This seems like a good change since reporting all errors is what dune tries to do by default.

Agreed that this is a positive change. It seems worth adding a test to specifically check that we report all "visible" errors, i.e. all errors that are not blocked by other errors, to avoid future regressions.

aalekseyev commented:
The changes look nice. 
Cutoff on `build_file_unit` seems dodgy, by the way. (It's equally dodgy before this PR, so this is merely a side discussion)
Isn't the code downstream from that accessing the file contents in a way that affects things (e.g. for dependency trace computation)?

jeremiedimino commented:
> @jeremiedimino I didn't understand the reasoning here. Why do you think adding a cutoff for this version isn't right?

It's expensive (need to hash the dep set), and right now the only call sites of this function ignore the output. Well, except for `evaluate_rules`, but we don't care about perf there.

> Agreed that this is a positive change. It seems worth adding a test to specifically check that we report all "visible" errors, i.e. all errors that are not blocked by other errors, to avoid future regressions.

Sounds good.

> Cutoff on build_file_unit seems dodgy, by the way. (It's equally dodgy before this PR, so this is merely a side discussion)
Isn't the code downstream from that accessing the file contents in a way that affects things (e.g. for dependency trace computation)?

Indeed. I guess we could make `build_file` return the file digest?

aalekseyev commented:
> I guess we could make `build_file` return the file digest?

Yeah, I think that can work.

jeremiedimino commented:
> Agreed that this is a positive change. It seems worth adding a test to specifically check that we report all "visible" errors, i.e. all errors that are not blocked by other errors, to avoid future regressions.

Done in #4309. I checked and the behaviour is improved by this PR.

snowleopard commented:
> Done in #4309. I checked and the behaviour is improved by this PR.

That's great, thanks!

jeremiedimino commented:
No problem. I rebased and updated the report-all-errors.t test, which now reports all the errors.

jeremiedimino commented:
> Yeah, I think that can work.

Alright, I'll that for a separate PR.

snowleopard reviewed (approved):
Looks good to me!
